debian/nadoka.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH nadoka "1" "Jul 2005"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
nadoka \- IRC Client Server program
.SH SYNOPSIS
.B nadoka
.RI --rc
.I rcfile
.RI [ options ]
.SH DESCRIPTION
This manual page documents briefly the
.B nadoka
commands.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBnadoka\fP is an IRC proxy program.  It always keeps to connect to IRC
server.
.PP
You can create bot program using nadoka easily.

.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-r, \-\-rc rcfile
Specify rc(config) file
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.SH SEE ALSO
nadoka has detailed document at following URL:
.br
.I http://www.atdot.net/nadoka/
.SH AUTHOR
nadoka was written by SASADA Koichi <ko1@atdot.net>.
.PP
This manual page was written by Taku YASUI <tach@debian.org>,
for the Debian project (but may be used by others).


ChangeLog
# ChangeLog of Nadoka($Rev: 210 $)
# $Id: ChangeLog 210 2009-09-16 19:27:37Z znz $ 

2009-09-17  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* plugins/googlebot.nb: use hl because lr used by web searcher only.

2009-09-17  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* plugins/googlebot.nb: support more searcher of Google AJAX
	Search API.

2009-09-17  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* plugins/googlebot.nb: use Google AJAX Search API.
	see http://code.google.com/intl/ja/apis/ajaxsearch/ for detail.

2009-09-14  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* plugins/googlebot.nb: Google SOAP Search API (No Longer Available).
	see http://code.google.com/intl/ja/apis/soapsearch/ and
	http://googlecode.blogspot.com/2009/08/well-earned-retirement-for-soap-search.html

2009-08-21  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* nadoka.rb, ndk/server.rb: revert previous change, and use ?\x1
	instead.

2009-08-21  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* nadoka.rb, ndk/server.rb: handle ctcp on ruby 1.9.

2009-08-18  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* nadoka.rb, ndk/config.rb, rice/irc.rb: ruby 1.9 support (not
	completed yet). apply patch from unak.

2009-08-17  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* nadokarc, ndk/config.rb, ndk/server.rb: no listen when
	Client_server_port is nil.

2009-08-17  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* plugins/tenkibot.nb: omit empty min celsius.

2009-07-29  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* lib/rss_check.rb: update lirs_uri.

2009-07-29  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* 0.7.1 : released

	* ndk/version.rb : 0.7.2

2009-07-27  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* nadoka.rb: add --daemon option. (apply patch from unak)

2009-07-04  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* plugins/dictbot.nb: avoid nkf bug.
	add waei as w, ruigo as r.
	new yahoo dict URI. (apply patch from ko1_ndk)

	* plugins/googlebot.nb: quote splited words. (apply patch from ko1_ndk)

2009-07-01  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* rice/irc.rb: @conn[0] may be nil.

2009-05-25  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* ndk/config.rb, ndk/server.rb: add Primitive_Filter
	  to filter any commands. (apply patch from unak)

2009-05-20  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* plugins/identifynickserv.nb: added.

2009-05-20  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* ndk/bot.rb, ndk/server.rb: add on_server_connected.

2009-04-16  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* nadokarc, ndk/config.rb, ndk/server.rb, rice/irc.rb: support SSL.

Wed Mar 19 19:46:52 2008  Kazuhiro NISHIYAMA  <zn@mbf.nifty.com>

	* ndk/logger.rb : pass ch to msgobj from clog.

	* ndk/config.rb : unify safe channel log filename.

Mon Jul 23 01:07:08 2007  Kazuhiro NISHIAYMA <zn@mbf.nifty.com>

	* lib/rss_check.rb : bug fix when dc_date is nil.

	* ndk/config.rb : fix bug when non-Hash in BotConfig array.


Mon Jul 02 19:01:04 2007  Koichi Sasada  <ko1@atdot.net>

	* plugins/tenkibot.nb : fix output format detail.

	* plugins/googlebot.nb : fix to retry 5 times if error is
          occurred while google search.

	* plugins/dictbot.nb : fix to show short summary.


2007-07-02(Mon) 18:01:58 +0900  Kazuhiro NISHIAYMA <zn@mbf.nifty.com>

	* rice/irc.rb : allow trailing "." in HOSTNAME (for NickServ
	on Freenode)


2007-02-20(Tue) 23:35:41 +0900  Kazuhiro NISHIAYMA <zn@mbf.nifty.com>

	* **/*.rb, plugins/*.nb : fix typos


2006-10-28(Sat) 22:08:11 +0900  Kazuhiro NISHIAYMA <zn@mbf.nifty.com>

	* ndk/server_state.rb : fix bug on_mode after server split


2006-10-20(Fri) 22:22:07 +0900  Kazuhiro NISHIAYMA <zn@mbf.nifty.com>

	* ndk/bot.rb : new method ccn2rcn


2006-10-07(Sat) 03:52:11 +0900  Koichi Sasada  <ko1@atdot.net>

	* plugin/tenkibot.nb : added


2006-10-02(Mon) 21:58:40 +0900  Koichi Sasada  <ko1@atdot.net>

	* 0.7.0 : released

	* ndk/version.rb : 0.7.1


2006-10-02(Mon) 21:41:24 +0900  Koichi Sasada  <ko1@atdot.net>

	* plugins/weatherbot.nb : removed

	* plugins/sendpingbot.nb : added

	* ndk/client.rb : modified for pingbot

	* ndk/server.rb : ditto


2006-09-01(Fri) 02:04:15 +0900  Kazuhiro NISHIAYMA <zn@mbf.nifty.com>

	* ndk/bot.rb : support send to a safe channel

	* nadokarc : fix misalignment


2006-08-31(Thu) 01:40:39 +0900  Kazuhiro NISHIAYMA <zn@mbf.nifty.com>

	* ndk/bot.rb : fix typos


2006-08-11(Fri) 14:45:01 +0900  Koichi Sasada  <ko1@atdot.net>

	* plugins/gonzuibot.nb : add a comma

	* plugins/pastebot.nb : add nick to paste page name

	* plugins/googlebot.nb : fix show_char_code_and_erace_tag


2006-05-16(Tue) 16:16:30 +0900  Koichi Sasada  <ko1@atdot.net>

	* ndk/server.rb : handle TERM signal
 

2006-03-27(Mon) 19:50:27 +0900  Koichi Sasada  <ko1@atdot.net>

	* plugins/dictbot.nb : fix dict index

	* plugins/roulettebot.nb : added

	* rice/irc.rb : loosen nick regexp

	* ndk/server.rb : randomize server if serching next server


2005-09-30(Fri) 00:05:19 +0900  Koichi Sasada  <ko1@atdot.net>

	* plugins/googlebot.nb : fix bugs

	* plugins/dictbot.nb : support

	* plugins/gonzuibot.nb : added

	* ndk/config.rb : fix bugs


2005-06-25(Sat) 19:02:24 +0900  Koichi Sasada  <ko1@atdot.net>

	* nadoka.rb : output error information  to file
	'nadoka_fatal_error' when unhandled error is occure (bug)


2005-06-14(Tue) 21:26:07 +0900  Koichi Sasada  <ko1@atdot.net>

	* ndk/server.rb : change reloaded message output method to slog


2005-06-13(Mon) 18:25:33 +0900  Koichi Sasada  <ko1@atdot.net>

	* nadokarc, ndk/logger.rb, ndk/config.rb : support
	:channel_name_in_file_name option for logging


2005-06-13(Mon) 00:11:39 +0900  Koichi Sasada  <ko1@atdot.net>

	* ndk/server_state.rb : fix typo

	* ndk/config.rb : see ENV['LANG'] to decide filename encoding

	* ndk/version.rb : fix version format

	* ndk/bot.rb : fix comment


2005-06-10(Fri) 04:38:25 +0900  Koichi Sasada  <ko1@atdot.net>

	* ndk/server.rb : fix logic operation


2005-06-09(Thu) 17:22:25 +0900  Koichi Sasada  <ko1@atdot.net>

	* ndk/logger.rb : synchronize IOLogWriter output



nadoka.rb
#
##
## Nadoka:
##  Irc Client Server Program
##
#
# Copyright (c) 2004-2005 SASADA Koichi <ko1 at atdot.net>
#
# This program is free software with ABSOLUTELY NO WARRANTY.
# You can re-distribute and/or modify this program under
# the same terms of the Ruby's license.
#
#
# $Id: nadoka.rb 206 2009-08-21 06:50:20Z znz $
# Create : K.S. 03/07/10 20:29:07
#

unless "".respond_to?(:force_encoding)
  class String
    def force_encoding(enc)
      self
    end
  end
end

$LOAD_PATH.unshift File.dirname(__FILE__)
require 'ndk/version'

if $0 == __FILE__

require 'optparse'

require 'ndk/server'
require 'ndk/bot'

$stdout.sync=true
$NDK_Debug  = false

unless defined? Process.daemon
  def Process.daemon(nochdir = nil, noclose = nil)
    exit!(0) if fork
    Process.setsid
    exit!(0) if fork
    Dir.chdir('/') unless nochdir
    File.umask(0)
    unless noclose
      STDIN.reopen('/dev/null')
      STDOUT.reopen('/dev/null', 'w')
      STDERR.reopen('/dev/null', 'w')
    end
  end
end

rcfile = nil
daemon = false
optparse = OptionParser.new{|opts|
  opts.banner = "Usage: ruby #{$0} [options]"

  opts.separator ""
  opts.separator "Require options:"

  opts.on("-r", "--rc [RCFILE]",
          "Specify rcfile(required)"){|f|
    rcfile = f
  }

  opts.separator ""
  opts.separator "Optional:"

  opts.on("-d", "--debug",
          "Debug Nadoka"){
    $NDK_Debug = true
    $DEBUG = true
    
    puts 'Enter Nadoka Debug mode'
  }
  opts.on("--daemon", "run as daemon"){
    daemon = true
  }

  opts.separator ""
  opts.separator "Common options:"

  opts.on_tail("-h", "--help", "Show this message"){
    puts Nadoka.version
    puts opts
    exit
  }
  opts.on_tail("-v", "--version", "Show version"){
    puts Nadoka.version
  }
}
optparse.parse!(ARGV)

unless rcfile
  puts Nadoka.version
  puts optparse
  exit
end

if daemon
  Process.daemon
end

begin
  GC.start
  Nadoka::NDK_Server.new(rcfile).start
rescue Nadoka::NDK_QuitProgram
  #
rescue Nadoka::NDK_RestartProgram
  retry
rescue Exception => e
  open('nadoka_fatal_error', 'w'){|f|
    f.puts e
    f.puts e.backtrace.join("\n")
  } 
end

end

